### PR TITLE
NET-581 - Added vault namespace in helm  

### PR DESCRIPTION
**Pull request by absolutelightning**
Changes proposed in this PR:
- Adds `vaultNamespace` in `secretsBackend.vault` in `values.yaml` 
- This namespace is used for Vault namespace, this introduces one more way to specify namespace other than
`"{"connect": [{ "ca_config": [{ "namespace": "value"}]}]}"` in `connectCA.additionalConfig`
- If `vaultNamespace` is present, it automatically adds annotation below to the templates.

```
    vault:
      agentAnnotations: |
        vault.hashicorp.com/namespace: vaultNamespace
```

How I've tested this PR:
a. CI
b. Updated test `TestVault_VaultNamespace`
Test Steps - 
```
1. kind create cluster --name=dc1
2. kind create cluster --name=dc2
3. cd acceptance/test/vault
4. go test ./... -p 1 -timeout 2h -failfast -no-cleanup-on-failure -debug-directory=/tmp/debug -use-kind -enable-multi-cluster -kube-contexts=kind-dc1,kind-dc2  -run ^TestVault
```
Output - 
```
asheshvidyut@absolutelightning-H2GX766V9T acceptance/tests (NET-581-Configure-Vault-namespaces-for-Connect-CA-via-Helm-Stanza) » go test ./... -p 1 -timeout 2h -failfast -no-cleanup-on-failure -debug-directory=/tmp/debug -use-kind -enable-multi-cluster -kube-contexts=kind-dc1,kind-dc2  -run ^TestVault
ok  	github.com/hashicorp/consul-k8s/acceptance/tests/api-gateway	0.492s [no tests to run]
ok  	github.com/hashicorp/consul-k8s/acceptance/tests/basic	0.547s [no tests to run]
ok  	github.com/hashicorp/consul-k8s/acceptance/tests/cli	0.529s [no tests to run]
ok  	github.com/hashicorp/consul-k8s/acceptance/tests/cloud	0.565s [no tests to run]
ok  	github.com/hashicorp/consul-k8s/acceptance/tests/config-entries	0.524s [no tests to run]
ok  	github.com/hashicorp/consul-k8s/acceptance/tests/connect	0.556s [no tests to run]
ok  	github.com/hashicorp/consul-k8s/acceptance/tests/consul-dns	0.543s [no tests to run]
ok  	github.com/hashicorp/consul-k8s/acceptance/tests/example	0.540s
ok  	github.com/hashicorp/consul-k8s/acceptance/tests/ingress-gateway	0.553s [no tests to run]
ok  	github.com/hashicorp/consul-k8s/acceptance/tests/metrics	0.555s [no tests to run]
ok  	github.com/hashicorp/consul-k8s/acceptance/tests/partitions	0.561s [no tests to run]
ok  	github.com/hashicorp/consul-k8s/acceptance/tests/peering	0.568s [no tests to run]
ok  	github.com/hashicorp/consul-k8s/acceptance/tests/sameness	0.552s
ok  	github.com/hashicorp/consul-k8s/acceptance/tests/snapshot-agent	0.572s [no tests to run]
ok  	github.com/hashicorp/consul-k8s/acceptance/tests/sync	0.554s [no tests to run]
ok  	github.com/hashicorp/consul-k8s/acceptance/tests/terminating-gateway	0.553s [no tests to run]
ok  	github.com/hashicorp/consul-k8s/acceptance/tests/vault	916.339s
ok  	github.com/hashicorp/consul-k8s/acceptance/tests/wan-federation	0.587s [no tests to run]
```

How I expect reviewers to test this PR:

Checklist:
- [x] Tests added
- [x] [CHANGELOG entry added](https://github.com/hashicorp/consul-k8s/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: acceptance/tests/vault/vault_namespaces_test.go
Comment (zalimeni) [inline acceptance/tests/vault/vault_namespaces_test.go:27]: Instead of copying the original test, could we refactor it into a helper that allows us to decide which form of config to use, and then run it once per case? That would reduce the risk of drift between the two tests in the future and make the difference between them more obvious.

File: charts/consul/values.yaml
Comment (zalimeni) [inline charts/consul/values.yaml:214]: (Outside comment range) Below, I think we'll want to remove the `"namespace"` KV from the `additionalConfig` example so as not to confuse operators about where it should be configured

File: charts/consul/values.yaml
Comment (zalimeni) [inline charts/consul/values.yaml:209]: ~ Align w/ other descriptions (if we keep this config key here)
```suggestion
        # Vault namespace (optional).
        # For more details, please refer to [Vault Connect CA configuration](https://developer.hashicorp.com/consul/docs/connect/ca/vault#namespace)
```

**Review by zalimeni — COMMENTED** [review]
Looks good to me overall! Added some minor feedback inline.

One clarifying question about requirements: I think @david-yu 's example in the ticket had `namespace` under the higher-level `secretsBackend.vault` key. I'm not opposed to keeping it under `connectCA` because that [aligns w/ the existing agent config](https://developer.hashicorp.com/consul/docs/connect/ca/vault#namespace), but I could see a higher-level key being used for other namespaced Vault config, if we expect that one Vault namespace would always be used across Vault->Consul integrations (I'm not 100% certain whether that's the case).

1. Do we think it's best to move the namespace KV up, or keep it under `connectCA`?
2. If we move it, a suggestion in one of the linked Slack 🧵s was to call it `vaultNamespace` to disambiguate from Consul. That would align w/ other names such as `consulServerRole`.

**Comment by david-yu** [comment]
@curtbushko Is there a reason why you only added backport 1.2.x instead of 1.0.x, 1.1.x and 1.2.x?

**Comment by curtbushko** [comment]
@david-yu No reason. I was just doing a passby label adding. I didn't look too deep to see where this should go.

**Comment by david-yu** [comment]
> -yu 's example in the ticket had `namespace` under the higher-level `secretsBackend.vault` key. I'm not opposed to keeping it under `connectCA` because that [aligns w/ the existing agent config](https://developer.hashicorp.com/consul/docs/connect/ca/vault#namespace), but I could see a higher-level key being used for other namespaced Vault config, if we expect that one Vault namespace would always be used across Vault->Consul integrations (I'm not 100% certain whether that's the case).
> 
> 1. Do we think it's best to move the namespace KV up, or keep it under `connectCA`?
> 2. If we move it, a suggestion in one of the linked Slack 🧵s was to call it `vaultNamespace` to disambiguate from Consul. That would align w/ other names such as `consulServerRole`.

I would prefer it be under the `secretsBackend.vault` stanza. If we do that we should automatically set agent annotations to the associated vault namespace as well as that is also required to get the vault namespace working. Below is an example of full desired secrets backend config to set the vault namespace to `admin`: 

```
vault:
      enabled: true
      vaultNamespace: admin
      consulServerRole: consul-server
      consulClientRole: consul-client
      consulCARole: consul-ca
      manageSystemACLsRole: consul-server-acl-init
      controllerRole: consul-controller
      connectInjectRole: consul-connect-inject
      controller:
        caCert:
          secretName: "consul_controller_int/cert/ca"
        tlsCert:
          secretName: "consul_controller_int/issue/consul-controller"
      connectInject:
        caCert:
          secretName: "consul_connect_inject_int/cert/ca"
        tlsCert:
          secretName: "consul_connect_inject_int/issue/consul-connect-inject"
      connectCA:
        address: https://hcp-vault-cluster:8200/
        rootPKIPath: consul_pki/
        intermediatePKIPath: consul_connect_int/
```

Behind the scenes it should also set the following annotation to `admin`:
```
    vault:
      agentAnnotations: |
        vault.hashicorp.com/namespace: admin
```

It would be best to call it `vaultNamespace` as well since it makes it clear this a vault namespace versus Consul.

**Comment by absolutelightning** [comment]
> 

@david-yu I am not super familiar with vault namespace. but can this situation arise where user would want different vault namespace for agentAnnotation and differnet for connectCA?

**Comment by david-yu** [comment]
I don't think it's very likely that the Connect CA will live in a different place than the Vault KV store but you can have it separate if they want to do that. I would think we set the namespace to the value provided by `vaultNamespace` by default and have it overridden by the user if need be.

File: charts/consul/values.yaml
Comment (absolutelightning) [reply charts/consul/values.yaml:215]: Do you mean remove from additional config?

**Comment by absolutelightning** [comment]
also can user give some other annotations as well for vault? we are going to remove agentAnnotations field right. 
Now it can only have one value. @david-yu

**Comment by absolutelightning** [comment]
@david-yu Removing agent annotations is probably incorrect as I see there are other annotations as well which users can add - https://developer.hashicorp.com/vault/docs/platform/k8s/injector/annotations. Let me know you thoughts.

**Comment by david-yu** [comment]
We shouldn't remove the agent annotations, mainly set them for you automatically if you set it via the `vaultNamespace` value. However they should also be overridden if the user wanted to.

File: charts/consul/values.yaml
Comment (absolutelightning) [inline charts/consul/values.yaml:233]: @zalimeni Did you intent to remove this - to avoid confusion for operators?

File: acceptance/tests/vault/vault_namespaces_test.go
Comment (absolutelightning) [inline acceptance/tests/vault/vault_namespaces_test.go:198]: Have kept only one test for namespace as it will take less time in CI and uses our new preferred way to set namespace for connectCA.

File: charts/consul/values.yaml
Comment (zalimeni) [inline charts/consul/values.yaml:118]: Since we're going w/ the higher-level config key, we should probably update this to also note the annotation it controls.
```suggestion
      # Vault namespace (optional). This sets the Vault namespace for the `vault.hashicorp.com/namespace` 
      # agent annotation and [Vault Connect CA configuration](https://developer.hashicorp.com/consul/docs/connect/ca/vault#namespace).
      # To override one of these values individually, see `agentAnnotations` and `connectCA.additionalConfig`.
```
@david-yu @absolutelightning does this look about right to you?

Also thinking we should test that both of these fields take precedence over the higher-level one we're adding when both are specified, if that's a use case we intend to support.

File: charts/consul/values.yaml
Comment (zalimeni) [reply charts/consul/values.yaml:233]: Yep! This seems no longer necessary to provide in the example because we now have an explicit field for it.

File: charts/consul/values.yaml
Comment (absolutelightning) [reply charts/consul/values.yaml:118]: @zalimeni Adding namespace to connectCA.additionalConfig will not update the annotation. It will just override the namespace. also Additional config is now not the preferred way so I had removed test for additional config.
Currently - higher-level one if present will add namespace to connectCA and create vault namespace annotation.
If annotations are already provided by user it will not append vault namespace annotation.
cc @david-yu is this correct?

Comment (david-yu) [reply charts/consul/values.yaml:118]: > @zalimeni Adding namespace to connectCA.additionalConfig will not update the annotation. It will just override the namespace. also Additional config is now not the preferred way so I had removed test for additional config.
> Currently - higher-level one if present will add namespace to connectCA and create vault namespace annotation.
> If annotations are already provided by user it will not append vault namespace annotation.
> cc @david-yu is this correct?

Yes that is what we should do. The docs do look correct to me.

Comment (zalimeni) [reply charts/consul/values.yaml:118]: > Adding namespace to connectCA.additionalConfig will not update the annotation. It will just override the namespace.

Yep, that makes sense 👍🏻 what I was trying to clarify in the docs is that there are two individual values that can be separately overridden, regardless of whether the top-level field is set.

> Additional config is now not the preferred way so I had removed test for additional config.

Based on updated docs proposed, I think that we'd want two specific test cases covered:
- if `vault.vaultNamespace` and `agentAnnotations` both specify a namespace, the `agentAnnotations` value will be used for the annotation
- if `vault.vaultNamespace` and `connectCA.additionalConfig` both specify a namespace, the `connectCA.additionalConfig` value will be used for the config

Comment (absolutelightning) [reply charts/consul/values.yaml:118]: ok yes.

Comment (absolutelightning) [reply charts/consul/values.yaml:118]: I have updated bats tests for point 1. but point 2 I am not able to achieve, not sure if bats accepts json in --set flag.
Can you please help. @zalimeni

Comment (absolutelightning) [reply charts/consul/values.yaml]: there is a -set-json flag but I think this will not be in bats tests. rather go tests.

Comment (absolutelightning) [reply charts/consul/values.yaml]: @zalimeni Point 1 is covered using bats tests.
Regarding Point 2 - do we want to keep this way of setting namespace in our codebase where we use additionalConfig to set namespace - we removed the comment [here](https://github.com/hashicorp/consul-k8s/pull/2841#discussion_r1310434027). Also i feel adding on more go test will increase CI time and since additionalConfig is no longer a preferred way we can exclude that. What are your thoughts?

File: acceptance/tests/vault/vault_namespaces_test.go
Comment (zalimeni) [inline acceptance/tests/vault/vault_namespaces_test.go:200]: ~ Should we remove this as well since it's now being covered by `vaultNamespace`?

File: charts/consul/templates/api-gateway-controller-deployment.yaml
Comment (zalimeni) [inline charts/consul/templates/api-gateway-controller-deployment.yaml:41]: Here and in other files:

Should this be an `else`? I might be wrong but I assumed you could specify `agentAnnotations` as well as `vaultNamespace` at the same time (annotations can include [many](https://developer.hashicorp.com/vault/docs/platform/k8s/injector/annotations#agent-annotations) values, not just namespace).

Seems this might also be lacking test coverage - maybe something we can add to the `correct vault namespace annotations is set when global.secretsBackend.vault.vaultNamespace is set and agentAnnotations are also set` test?

File: charts/consul/values.yaml
Comment (zalimeni) [reply charts/consul/values.yaml]: @absolutelightning I'm not sure how much of a priority it is to preserve the option of using `additionalConfig` namespace as an override for the new KV - I'm more just thinking if we document it, we should confirm it works. An alternative would be not to document it for now, if it's difficult to test and we're not sure it's going to be needed by customers. Maybe @david-yu has thoughts?

Re: docs (if we do keep it), I don't feel strongly about removing the example, so happy to defer to you and David on that. I agree it's more nuanced in the override case; it was a simpler decision when we were just adding an explicit field for `connectCA`.

File: charts/consul/templates/api-gateway-controller-deployment.yaml
Comment (zalimeni) [reply charts/consul/templates/api-gateway-controller-deployment.yaml:41]: @absolutelightning if I set an unrelated annotation (not namespace) via `agentAnnotations`, wouldn't we still expect setting `vaultNamespace` to add the namespace annotation? My read of this is that we have made the two config values mutually exclusive, even if `agentAnnotations` does not contain `vaultNamespace`.

Comment (absolutelightning) [reply charts/consul/templates/api-gateway-controller-deployment.yaml:41]: According to me we will not know if namespace annotation is present or not in agent annotations added by user. So Ideally we would check if namespace annotation is present or not in agent annotations and then add namespace annotation in case user sets vaultNamespace. But I am not sure if it is possible.

Comment (absolutelightning) [reply charts/consul/templates/api-gateway-controller-deployment.yaml:41]: yes.

Comment (absolutelightning) [reply charts/consul/templates/api-gateway-controller-deployment.yaml]: > > Ideally we would check if namespace annotation is present or not in agent annotations and then add namespace annotation in case user sets vaultNamespace.
> 
> Here is my thinking of what I expect in the following cases:
> 
> 1. `vaultNamespace` set and no namespace annotation in `agentAnnotations` - add the agent annotations to the existing annotations the user provides
> 2. No `vaultNamespace` set and `agentAnnotations` namespace set - Just use the agentAnntotations that have been set which includes that namespace annotation which the user manually set.
> 3. `vaultNamespace` set and namespace annotation in `agentAnnotations` set - Leave annotations as is. We'd bet setting the ConnectCA in one namespace and set agent annotations for the vault agent with another for KV storage.
> 4. `vaultNamespace` not set and no `agentAnnotations` set - Do nothing.
> 
> I think I covered all permutations?

All cases are handled in bats test @zalimeni

File: charts/consul/templates/server-config-configmap.yaml
Comment (zalimeni) [inline charts/consul/templates/server-config-configmap.yaml:77]: I think we would want to check `additionalConfig` similar to the annotations so we don't duplicate namespace values.

I think I figured out how to do actual key checks rather than `contains` in the embedded JSON and YAML, [here's](https://helm-playground.com/#t=LQhQEMDtIewF3HAljSBnASgUzQVwDZwBcoA3qcAARIBmlAFFACYMB0AauPrjqwOb4YAIy6s0WAMYAnLHDQAhcBIDWWSE1YA3cAThadhAHLgAtjgAOSrAEoGsOAwAW4NAGksATwZMsNAw4AiAMoOLh40fkERfDFJGTlFFTUNbV1WcD41OABBaHhEFHRKAB9KGikYEwBNU3xbANTCVmc0RyQJGClzVg6TAHpIUwsrAOsxgF9x0EpKBv9mlzaOrp7KgaG0SwksAKJZ8hDObl4BYVFxaVkFJVV1fTTGuGMzTatKSYCyCkpk96nQDqQGhIPjYPCEEgzchUADuSDgjkOYROUXOcSuiVuKX8f2mlEBkEkcGAEnAJNQwL4rAAVmhUHtinioUyZrMCUTdpQANos1mUUi8vls8AAfUBlM5PKF0v5gpls3ATCYMjQaE5AQOq2gRIAwtl0kqVWg-gEADRymXQ6h0ZghUmxS5PIZ28AO%2BLuLyTC3SgKkkXA-A7PYBPqA8F9R4k8B9OD4CLSOBm71Cq2-L3yn1ISBwLBSMxMJCILAi8zKJAlxCOdWa9kSOB61hZnN5rAFosABVcAEl25WTeaMynvrQGLb6PcjBstjY7PAnC4Pd5fDignbUIS6w3FQXkKguDqKSCSmUKiYAFJ0yD1QYvaejMa4wd8gI34bbaukEKPZ5vrD95N8qm6iPk%2BMwBBU8AlmWFYIh%2Ba7apu%2BoQXAnY9n2HwDqBCq4AiIpmAiMBMJyApYayARwB45hBrMyi4EIuaEjmaqYaRAQmDAuDZjBVbBjW666vqOgIgAsrIjiEb2CL-qRYGWFIphqnsJEyWBFSBnBWrggAyrmmi5hgMCBiaAFClMpFmU%2BFnygAuixMq%2BqK5gVJoSA%2BFI6qPJ88pWay1kshZQEsJMQA&v=LQhQHMBsHsCMENIC5QAJUGcCmBjATlgC4YBC8OA1lgHYAmK66AbvAK6SEOPNscBy8ALZYMAB3JYkqAESFoo4JCxMskYNQzS03VPHA1CAQWrVoheIQCW0DVIA%2B2nahbtCAOgAW8DB8s5oeKJu-oIA9NRCIuI4kqgAZtDQwPAmZhbW1I46Lhye3r7%2BgcHQYWYeWHhSCHhZ6P4muIQAwoZcTvC0tJZWNohNNnGW4Pa1OgDeqKNOMhHCYhLSUtIJSQBWGDbSADRTTtJlFYsy1VrTjAC%2BoEA) an example.

Something like that should allow us to have clean mutual exclusion on both config entries, w/ the more specific values taking precedence over `vault.vaultNamespace`, I think. In this case, it would look like:

```suggestion
              {{- if (and (.vaultNamespace) (not (hasKey (default "" .connectCA.additionalConfig | fromJson) "namespace"))) }}
              "namespace": "{{ .vaultNamespace }}",
              {{- end }}
```

For the annotation checks, I think we can do:
```
{{- if (and (.Values.global.secretsBackend.vault.vaultNamespace) (not (hasKey (default "" .Values.global.secretsBackend.vault.agentAnnotations | fromYaml) "vault.hashicorp.com/namespace")))}}
  "vault.hashicorp.com/namespace": "{{ .Values.global.secretsBackend.vault.vaultNamespace }}"
{{- end }}
```

File: charts/consul/test/unit/server-config-configmap.bats
Comment (zalimeni) [inline charts/consul/test/unit/server-config-configmap.bats:661]: Re: other comment, can we also check here for conditional overrides when both are set?

**Comment by zalimeni** [comment]
One last bit of feedback re: opportunity to do better detection of conflicts between config values. Otherwise, this is looking great to me! LMK when you're ready for another look and I'll try to be speedy. Thanks for your work on this, @absolutelightning 🙏🏻

File: .changelog/2841.txt
Comment (david-yu) [inline .changelog/2841.txt:2]: This likely needs to change based on the new values we decided from the review.

File: charts/consul/templates/server-config-configmap.yaml
Comment (absolutelightning) [inline charts/consul/templates/server-config-configmap.yaml:75]: json structure is like this https://github.com/hashicorp/consul-k8s/pull/2841/files#diff-73787600ab1d7b64c9a865fad8d8520d230644c998b3bf9b5670ffdb8900bb2eR713.
was having trouble fetching keys from nested array of json. hence had to keep it as string.
@zalimeni

File: .changelog/2841.txt
Comment (david-yu) [inline .changelog/2841.txt:2]:
```suggestion
vault: Adds `secretsBackend.vault.vaultNamespace` to configure a new value `secretsBackend.vault.connectCA` in Helm chart and apply the following value "vault.hashicorp.com/namespace: namespace" to `secretsBackend.vault.agentAnnotations` 
```

**Comment by david-yu** [comment]
Added consul-docs for review of values.yaml

File: charts/consul/test/unit/server-config-configmap.bats
Comment (zalimeni) [inline charts/consul/test/unit/server-config-configmap.bats:699]: This test case title looks off to me based on what it's testing. Can you verify this is what was intended?
```suggestion
@test "server/ConfigMap: do not set Vault Namespace in connect CA config from global.secretsBackend.vault.vaultNamespace when also set in connectCA.additionalConfig" {
```

File: charts/consul/templates/server-config-configmap.yaml
Comment (absolutelightning) [reply charts/consul/templates/server-config-configmap.yaml:75]: Screenshot 2023-09-06 at 10 52 32 AM
This is not the correct syntax. 
See - https://github.com/hashicorp/consul-k8s/blob/8f52b7c98e6f5b761c0ac69798c28ae152178de1/charts/consul/values.yaml#L234-L242
its inside json array.

Comment (zalimeni) [reply charts/consul/templates/server-config-configmap.yaml:75]: @absolutelightning ah, I'm sorry - that was unintuitive to me and I missed the `[]`s this whole time. (I also suspect looking at Consul code on the server side that it's not strictly necessary and this is an example of our JSON->HCL parsing complexity... 😞)

This looks overly complicated to detect in Helm template syntax even w/ the `first` template function as a sort-of option, so I feel a `"namespace"` `contains` check is a reasonable trade-off, given how unlikely it is someone would use that as a value (which would cause us to skip the `vaultNamespace` addition).

cc @thisisnotashwin in case you have any other thoughts, since we discussed this before.

File: .changelog/2841.txt
Comment (zalimeni) [inline .changelog/2841.txt:4]: ~ Optional, thought might help to simplify this entry based on recent changes:
```suggestion
vault: Adds `secretsBackend.vault.vaultNamespace` to Helm chart. If set, the value will be used in the `vault.hashicorp.com/namespace` annotation under `secretsBackend.vault.agentAnnotations` and the `namespace` field in `connectCA.additionalConfig`, if not already configured by those fields.
This provides a more convenient way to specify a single value for use across multiple fields that configure a Vault namespace.
```

**Review by zalimeni — APPROVED** [review]
LGTM pending any input from docs team or @david-yu.

**Review by david-yu — APPROVED** [review]

**Comment by absolutelightning** [comment]
> LGTM pending any input from docs team or @david-yu. Thank you again for working through my feedback and questions, @absolutelightning !

Thanks @zalimeni for review.

**Comment by absolutelightning** [comment]
@zalimeni I am not able to merge. can you please help?

**Pull request merged**
